packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java

```diff
@@ -0,0 +1,214 @@
+    if (broadcastReceiver == null) return;
+    activity.unregisterReceiver(broadcastReceiver);
+    broadcastReceiver = null;
+  }
+
+  private boolean isSystemAutoRotationLocked() {
+    return android.provider.Settings.System.getInt(
+            activity.getContentResolver(), Settings.System.ACCELEROMETER_ROTATION, 0)
+        != 1;
+  }
+
+  private PlatformChannel.DeviceOrientation getUIOrientation() {
+    final int rotation = getDisplay().getRotation();
+    final int orientation = activity.getResources().getConfiguration().orientation;
+
+    switch (orientation) {
+      case Configuration.ORIENTATION_PORTRAIT:
+        if (rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_90) {
+          return PlatformChannel.DeviceOrientation.PORTRAIT_UP;
+        } else {
+          return PlatformChannel.DeviceOrientation.PORTRAIT_DOWN;
+        }
+      case Configuration.ORIENTATION_LANDSCAPE:
+        if (rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_90) {
+          return PlatformChannel.DeviceOrientation.LANDSCAPE_LEFT;
+        } else {
+          return PlatformChannel.DeviceOrientation.LANDSCAPE_RIGHT;
+        }
+      default:
+        return PlatformChannel.DeviceOrientation.PORTRAIT_UP;
+    }
+  }
+
+  private PlatformChannel.DeviceOrientation calculateSensorOrientation(int angle) {
+    final int tolerance = 45;
+    angle += tolerance;
+
+    // Orientation is 0 in the default orientation mode. This is portait-mode for phones
+    // and landscape for tablets. We have to compensate for this by calculating the default
+    // orientation, and apply an offset accordingly.
+    int defaultDeviceOrientation = getDeviceDefaultOrientation();
+    if (defaultDeviceOrientation == Configuration.ORIENTATION_LANDSCAPE) {
+      angle += 90;
+    }
+    // Determine the orientation
+    angle = angle % 360;
+    return new PlatformChannel.DeviceOrientation[] {
+          PlatformChannel.DeviceOrientation.PORTRAIT_UP,
+          PlatformChannel.DeviceOrientation.LANDSCAPE_LEFT,
+          PlatformChannel.DeviceOrientation.PORTRAIT_DOWN,
+          PlatformChannel.DeviceOrientation.LANDSCAPE_RIGHT,
+        }
+        [angle / 90];
+  }
+
+  private int getDeviceDefaultOrientation() {
+    Configuration config = activity.getResources().getConfiguration();
+    int rotation = getDisplay().getRotation();
+    if (((rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_180)
+            && config.orientation == Configuration.ORIENTATION_LANDSCAPE)
+        || ((rotation == Surface.ROTATION_90 || rotation == Surface.ROTATION_270)
+            && config.orientation == Configuration.ORIENTATION_PORTRAIT)) {
+      return Configuration.ORIENTATION_LANDSCAPE;
+    } else {
+      return Configuration.ORIENTATION_PORTRAIT;
+    }
+  }
+
+  @SuppressWarnings("deprecation")
+  private Display getDisplay() {
+    return ((WindowManager) activity.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
+  }
+}
```

Comment:
This class has a lot of non-trivial logic, but no tests.

Reply:
Done.